File: coding-interviewes/coding_interview_150/java/JumpGame.java

/* Jump Game — greedy max reach. O(n) time, O(1) space */
public class JumpGame {
    public static boolean canJump(int[] nums){
        int reach=0;
        for (int i=0;i<nums.length;i++){ if (i>reach) return false; reach=Math.max(reach, i+nums[i]); }
        return true;
    }
    public static void main(String[] args){
        System.out.println(canJump(new int[]{2,3,1,1,4}));
        System.out.println(canJump(new int[]{3,2,1,0,4}));
    }
}

File: coding-interviewes/coding_interview_150/scala/JumpGame.scala

/* Jump Game — greedy max reach. O(n) */
object JumpGame {
  def canJump(nums:Array[Int]):Boolean = {
    var reach=0
    for (i <- 0 until nums.length){
      if (i>reach) return false
      reach = Math.max(reach, i + nums(i))
    }
    true
  }
  def main(args:Array[String]):Unit = {
    println(canJump(Array(2,3,1,1,4)))
    println(canJump(Array(3,2,1,0,4)))
  }
}
